src/PowerPing/Ping.cs

```diff
@@ -415,38 +399,54 @@ private PingResults SendICMP(PingAttributes attrs)
                     }
 
                     ICMP response;
-                    TimeSpan replyTime;
+                    EndPoint responseEP = iep;
+                    TimeSpan replyTime = TimeSpan.Zero;
                     do {
+                        // Cancel if requested
+                        cancellationToken.ThrowIfCancellationRequested();
+
+                        // Set receive timeout
+                        int remainingTimeout = (int)Math.Ceiling(attrs.Timeout - replyTime.TotalMilliseconds);
+                        if (remainingTimeout <= 0) {
+                            throw new SocketException();
+                        }
+                        sock.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReceiveTimeout, Math.Min(remainingTimeout, 250));
+
                         // Wait for response
-                        bytesRead = sock.ReceiveFrom(receiveBuffer, ref ep);
+                        try {
+                            bytesRead = sock.ReceiveFrom(receiveBuffer, ref responseEP);
+                        } catch (SocketException) {
+                            bytesRead = 0;
+                        }
                         replyTime = new TimeSpan(Helper.StopwatchToTimeSpanTicks(Stopwatch.GetTimestamp() - requestTimestamp));
 
-                        // Store reply packet
-                        response = new ICMP(receiveBuffer, bytesRead);
-
-                        // Ignore unexpected echo responses
-                        if (packet.type == 8 && response.type == 0) {
-                            ushort responseSessionId = BitConverter.ToUInt16(response.message, 0);
-                            ushort responseSequenceNum = BitConverter.ToUInt16(response.message, 2);
-                            if (responseSessionId != sessionId || responseSequenceNum != sequenceNum) {
-                                if (replyTime.TotalMilliseconds >= attrs.Timeout) {
-                                    throw new SocketException();
+                        if (bytesRead == 0) {
+                            response = null;
+                        } else {
+                            // Store reply packet
+                            response = new ICMP(receiveBuffer, bytesRead);
+
+                            // Ignore unexpected echo responses
```

Comment:
If feel like we should clarify that we are ignoring unexpected echo responses under normal sending conditions.